lighthouse-core/gather/driver/prepare.js

```diff
@@ -0,0 +1,177 @@
+const emulation = require('../../lib/emulation.js');
+const pageFunctions = require('../../lib/page-functions.js');
+
+/**
+ * Enables `Debugger` domain to receive async stacktrace information on network request initiators.
+ * This is critical for tracking attribution of tasks and performance simulation accuracy.
+ * @param {LH.Gatherer.FRProtocolSession} session
+ */
+async function enableAsyncStacks(session) {
+  const enable = async () => {
+    await session.sendCommand('Debugger.enable');
+    await session.sendCommand('Debugger.setSkipAllPauses', {skip: true});
+    await session.sendCommand('Debugger.setAsyncCallStackDepth', {maxDepth: 8});
+  };
+
+  // Resume any pauses that make it through `setSkipAllPauses`
+  session.on('Debugger.paused', () => session.sendCommand('Debugger.resume'));
+
+  // `Debugger.setSkipAllPauses` is reset after every navigation, so retrigger it on main frame navigations.
+  // See https://bugs.chromium.org/p/chromium/issues/detail?id=990945&q=setSkipAllPauses&can=2
+  session.on('Page.frameNavigated', event => {
+    if (event.frame.parentId) return;
+    enable().catch(err => log.error('Driver', err));
+  });
+
+  await enable();
+}
+
+/**
+ * Use a RequestIdleCallback shim for tests run with simulated throttling, so that the deadline can be used without
+ * a penalty.
+ * @param {LH.Gatherer.FRTransitionalDriver} driver
+ * @param {LH.Config.Settings} settings
+ * @return {Promise<void>}
+ */
+async function shimRequestIdleCallbackOnNewDocument(driver, settings) {
+  await driver.executionContext.evaluateOnNewDocument(pageFunctions.wrapRequestIdleCallback, {
+    args: [settings.throttling.cpuSlowdownMultiplier],
+  });
+}
+
+/**
+ * Dismiss JavaScript dialogs (alert, confirm, prompt), providing a
+ * generic promptText in case the dialog is a prompt.
+ * @param {LH.Gatherer.FRProtocolSession} session
+ * @return {Promise<void>}
+ */
+async function dismissJavaScriptDialogs(session) {
+  session.on('Page.javascriptDialogOpening', data => {
+    log.warn('Driver', `${data.type} dialog opened by the page automatically suppressed.`);
+
+    session
+      .sendCommand('Page.handleJavaScriptDialog', {
+        accept: true,
+        promptText: 'Lighthouse prompt response',
+      })
+      .catch(err => log.warn('Driver', err));
+  });
+
+  await session.sendCommand('Page.enable');
+}
+
+/**
+ * @param {LH.Gatherer.FRProtocolSession} session
+ * @param {{url: string}} navigation
+ * @return {Promise<{warnings: Array<LH.IcuMessage>}>}
+ */
+async function resetStorageForNavigation(session, navigation) {
+  /** @type {Array<LH.IcuMessage>} */
+  const warnings = [];
+
+  // Reset the storage and warn if there appears to be other important data.
+  const warning = await storage.getImportantStorageWarning(session, navigation.url);
+  if (warning) warnings.push(warning);
+  await storage.clearDataForOrigin(session, navigation.url);
+  await storage.clearBrowserCaches(session);
```

Comment:
it's new that these two `clear` are colocated. I couldn't think of a compelling reason why you'd only want to clear the data but not the caches, and we previously only cleared the caches on the same passes as the ones we cleared data so from protocol message log perspective it should behave the same as before.